TreeList Search Panel: rewrite (#781)
DevExpress/devextreme-demos: JSDemos/Demos/TreeList/UsingSearchPanel/description.md

> The search panel allows users to filter data by all columns. To enable this feature, set **searchPanel**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#visible) to **true**. If the TreeList should not search through a specific column, set its [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowSearch) property to **false**.

Reviewer: I moved the information about `allowSearch` to the next paragraph.

```suggestion
If you want users to search data, display the [searchPanel](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/). Set its [visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#visible) property to **true**.
```

> Search is case-insensitive. The default configuration is to search by [visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) and invisible columns. If you want to search by visible columns only, set the **searchPanel**.[searchVisibleColumnsOnly](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#searchVisibleColumnsOnly) property to **true**.

Reviewer: "The default configuration is to search" - Don't follow "the default behavior" or "the default configuration" with "is to do X". This reads poorly.

"Search is case-insensitive." - The paragraph discusses excluding columns from search, and this sentence isn't relevant. I moved it to the next paragraph, where we discuss how search applies to different data types.

```suggestion
The TreeList searches in all columns, regardless of whether they are [visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) or hidden. To exclude hidden columns from search, enable the **searchPanel**.[searchVisibleColumnsOnly](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#searchVisibleColumnsOnly) property. You can also exclude any specific column. To do so, set its [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowSearch) property to **false**
```

> Column's [data type](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#dataType) determines the condition under which the TreeList includes a row into search results:
> 
> * *"number"*, *"boolean"*, and *"data"* column values should be strictly equal to the search query
> 
> * *"string"* column values should contain the search query.

Reviewer: Lists are great when you need to enumerate similar things. This is not the case. Here, we are drawing a correspondence between two things. A table could fit in this case, but there is so little information to convey. I think we can go with a simple paragraph:

```suggestion
Search is case-insensitive. Numeric, date, and Boolean values match only if they are equal to the search query. String values match even if they contain only a part of the query.
```

> You can populate search panel with an initial query. To do this, specify the **searchPanel**.[text](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#text) property. You can also call the [searchByText(text)](/Documentation/ApiReference/UI_Components/dxTreeList/Methods/#searchByTexttext) method to specify query at runtime. 

Reviewer: ```suggestion
To specify a search query in code, set the **searchPanel**.[text](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#text) property or call the [searchByText(text)](/Documentation/ApiReference/UI_Components/dxTreeList/Methods/#searchByTexttext) method.
```

> The TreeList displays matched rows with their ancestors. For other display modes, see the [Filter Modes](https://js.devexpress.com/Demos/WidgetsGallery/Demo/TreeList/FilterModes) demo.

Reviewer: ```suggestion
```

> The search panel allows users to filter data by all columns: if at least one row field matches the query, the TreeList will display the corresponding record with its ancestors. To enable this feature, set **searchPanel**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#visible) to **true**.

Reviewer 2: I think that we can be more straightforward here:

```suggestion
The search panel allows users to filter data by all columns. If at least one row field matches the query, the TreeList displays such row with its ancestors. To enable this feature, set **searchPanel**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#visible) to **true**.
```

> The default behavior is to search by visible and invisible columns (**columns[]**.][visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible)). If you want to search by visible columns only, set the [searchVisibleColumnsOnly](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#searchVisibleColumnsOnly) property to **true**.

Reviewer 2: ```suggestion
The default behavior is to search by [visible and invisible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) columns. If you want to search by visible columns only, set the **searchPanel**.[searchVisibleColumnsOnly](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#searchVisibleColumnsOnly) property to **true**.
```

If you want users to search data, display the [searchPanel](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/). Set its [visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#visible) property to **true**.

> The TreeList searches in all columns, regardless of whether they are [visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) or hidden. To exclude hidden columns from search, enable the **searchPanel**.[searchVisibleColumnsOnly](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#searchVisibleColumnsOnly) property. You can also exclude any specific column. To do so, set its [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowSearch) property to **false**

Reviewer 3: ```suggestion
The TreeList searches in all columns, regardless of whether they are [visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) or hidden. To exclude hidden columns from search, enable the **searchPanel**.[searchVisibleColumnsOnly](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/searchPanel/#searchVisibleColumnsOnly) property. You can also exclude any specific column. To do this, set its [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowSearch) property to **false**
```

> Search is case-insensitive. Numeric, date, and Boolean values match only if they are equal to the search query. String values match even if they contain only a part of the query.

Reviewer 4: 1. I'd have the sentences about "case insensitive" and "string values" together, not separated. Most likely, I'd move the "case-insensitive" sentence to the end. It doesn't apply to numeric/date-time/Boolean values anyway, right? 
2. If I'm looking for "12" and a numeric column contains the number 123, it's not a match? That's what the sentence says, I'm just making sure. Or I can imagine that I want to find all dates in December and my search string is "12/". But as I understand, the search will not find any date-time values if I specify such a string? 
3. The last sentence - I'm also not sure about it. What you say is that I can enter "Vladimir". A part of the query is "Vlad". If a cell value contains the string "Vladislav", then it does contain "a part of the query", which is "Vlad". So I enter Vladimir and Vladislav is a match? Sounds strange.

Author: 1) "case-insensitive" is only for strings. Will move this sentence.
2) Right, search will not find any date-time values for such strings. They should be equal to, not contain the search query.
3) I mean that the query "ab" will list "abcd" and "abzy" values. "String values match even if they are partially equal to the search query."?

Reviewer 4: 1. Thanks
2. Alright - I was just making sure. But then the following question arises. Obviously dates or numbers can be formatted differently based on locale or maybe column settings. It could be 1 or $1.00. Or it could be 12/31/2020 or 31.12.2020. You say "values match" ... and I'm not sure which of those options are "values". The cell text needs to match the query or actually some underlying value? And if I show check boxes for the Boolean column, can I still search for "true"? Or "True"? 
3. OK - that's clearer, thanks. I'd probably say "contains". "String values match if they contain the search query".

Author: 2) If column's type is "currency" (e.g. the displayed value is "$2"), then the queries "$2" and "2" will list the corresponding row.
The search for dates works when the date is full (12/31/2020). It also works when a column uses a custom format ({ type: "longDate" }), but the query should be "Sunday, October 10, 2020". For custom scenarios, developers have to implement calculateFilterExpression. 
If column's type is "boolean", rows will be shown when the query is "true" (not "True"). We've decided to leave it as it is.
